src/Lecture2.hs

```diff
@@ -152,12 +183,124 @@ You're free to define any helper functions.
+            | otherwise                          = go (strikeNumber + 1) (reduceEndurance knight') (damageDragon knight' dragon)
+          fightResult :: (Knight, EvilDragon)
+          fightResult = go 1 knight evilDragon
+
+maxStrikesInRound :: Int
+maxStrikesInRound = 10
+
+evaluateResult :: EvilDragon -> Knight -> Either Error FightResult
+evaluateResult evilDragon knight
+  | isDead (dragonHealth dragon) = Right $ KnightWin $ createReward evilDragon
+  | isEnduranceEmpty (knightEndurance knight) = Right KnightRunAway
+  | isDead (knightHealth knight) = Right KnightDie
+  | otherwise = Left $ Error "something went wrong :("
+  where dragon = extractDragon evilDragon
+
+isFightFinished :: EvilDragon -> Knight -> Bool
+isFightFinished dragon knight = isDead (knightHealth knight)
+                             || isDead (dragonHealth (extractDragon dragon))
+                             || isEnduranceEmpty (knightEndurance knight)
+
+isDead :: Health -> Bool
+isDead (Health h) = h <= 0
+
+damageKnight :: EvilDragon -> Knight -> Knight
+damageKnight dragon knight = knight {knightHealth = reduceHealthByFirePower (knightHealth knight) (dragonFirePower (extractDragon dragon))}
+
+damageDragon :: Knight -> EvilDragon -> EvilDragon
+damageDragon knight evilDragon = embedDragon (dragon {dragonHealth = reduceHealthByAttack (dragonHealth dragon) (knightAttack knight)}) evilDragon
+  where dragon = extractDragon evilDragon
+
+reduceEndurance :: Knight -> Knight
+reduceEndurance knight@(Knight _ _ (Endurance e)) = knight {knightEndurance = Endurance (e - 1)}
+
+createReward :: EvilDragon -> Reward
+createReward dragon = let chest = extractChest dragon
+                          gold' = gold chest
+                          threasure' = threasure chest
+                          xp = extractXP dragon
+                      in Reward gold' threasure' xp
+
+extractDragon :: EvilDragon -> Dragon
+extractDragon (Green d _) = d
+extractDragon (Black d _) = d
+extractDragon (Red d _)   = d
+
+embedDragon :: Dragon -> EvilDragon -> EvilDragon
+embedDragon dragon (Green _ gold)  = Green dragon gold
+embedDragon dragon (Black _ chest) = Black dragon chest
+embedDragon dragon (Red _ chest)   = Red dragon chest
+
+extractChest :: EvilDragon -> Chest Treasure
+extractChest (Green _ g) = Chest g Nothing
+extractChest (Black _ c) = c
+extractChest (Red _ c)   = c
+
+extractXP :: EvilDragon -> XP
+extractXP (Green _ _) = XP 250
+extractXP (Black _ _) = XP 150
+extractXP (Red _ _)   = XP 100
+
+isEnduranceEmpty :: Endurance -> Bool
+isEnduranceEmpty (Endurance e) = e <= 0
+
+reduceHealthByFirePower :: Health -> FirePower -> Health
+reduceHealthByFirePower (Health h) (FirePower fp) = Health (h - fp)
+
+reduceHealthByAttack :: Health -> Attack -> Health
+reduceHealthByAttack (Health h) (Attack a) = Health (h - a)
+
+-- Q1: What is supposed to do to avoid part of this boilerplate functions? Optics?
```

Comment:
What is the specific boilerplate you're referring to? 🤔

Reply:
for a moment I thought I was writing too many functions just to unwrap newtypes values. I guess it's fine after all

Reply:
That's fine, especially if you work with different newtypes 👍🏻 

You can define `newtype`s as records as well if you only need extracting functions. And on Lecture 3 I'm going to explain `deriving` which can reduce some boilerplate around newtypes.